sunpy/util/net.py

```diff
@@ -111,6 +111,159 @@ def download_file(url, directory, default=u'file', overwrite=False):
+                        if url_key.startswith(cur_url):
+                            url_mirror = url_key.replace(cur_url,
+                                                         conf.dataurl_mirror)
+                            if url_mirror in url2hash:
+                                return url2hash[url_mirror]
+        with urllib.request.urlopen(remote_url, timeout=timeout) as remote:
+            info = remote.info()
+            if 'Content-Length' in info:
+                try:
+                    size = int(info['Content-Length'])
+                except ValueError:
+                    size = None
+            else:
+                size = None
+
+            if size is not None:
+                check_free_space_in_dir(directory, size)
+
+            if show_progress:
+                progress_stream = sys.stdout
+            else:
+                progress_stream = io.StringIO()
+
+            dlmsg = "Downloading {0}".format(remote_url)
+            with ProgressBarOrSpinner(size, dlmsg, file=progress_stream) as p:
+                if not overwrite and os.path.exists(path):
+                    path = replacement_filename(path)
+                with open(path, 'wb') as f:
+                    try:
+                        bytes_read = 0
+                        block = remote.read(conf.download_block_size)
+                        while block:
+                            f.write(block)
+                            bytes_read += len(block)
+                            p.update(bytes_read)
+                            block = remote.read(conf.download_block_size)
+                    except BaseException:
+                        if os.path.exists(f.name):
+                            os.remove(f.name)
+                        raise
+
+        if cache:
+            _acquire_download_cache_lock()
+            try:
+                with shelve.open(urlmapfn) as url2hash:
+                    # We check now to see if another process has
+                    # inadvertently written the file underneath us
+                    # already
+                    if url_key in url2hash:
+                        return url2hash[url_key]
+                    url2hash[url_key] = f.name
+            finally:
+                _release_download_cache_lock()
+
+    except urllib.error.URLError as e:
+        if hasattr(e, 'reason') and hasattr(e.reason, 'errno') and e.reason.errno == 8:
+            e.reason.strerror = e.reason.strerror + '. requested URL: ' + remote_url
+            e.reason.args = (e.reason.errno, e.reason.strerror)
+        raise e
+    except socket.timeout as e:
+        # this isn't supposed to happen, but occasionally a socket.timeout gets
+        # through.  It's supposed to be caught in `urrlib2` and raised in this
+        # way, but for some reason in mysterious circumstances it doesn't. So
+        # we'll just re-raise it here instead
+        raise urllib.error.URLError(e)
+    return f.name
```

Comment:
What are the differences between this and the astropy version?

Reply:
The differences here are that
1. The filename is not the name of hash, but that of the file downloaded. [here](https://github.com/vermashresth/sunpy/blob/issue2088/sunpy/util/net.py#L178:L180)
```
    opn = urlopen(remote_url)
    filename = get_system_filename(opn, remote_url, default)
    path = os.path.join(directory, filename.decode('utf-8'))
```
2. If file was downloaded earlier but new path is provided and `cache=True`, then file is just moved from old location to new one and mapping list is updated. [here](https://github.com/vermashresth/sunpy/blob/issue2088/sunpy/util/net.py#L192:L194)
```
    if url_key in url2hash:
                    if url2hash[url_key] != path:
                        shutil.move(url2hash[url_key], path)
                        url2hash[url_key] = path
                    return url2hash[url_key]
```
3. Optional overwrite parameter added, which was in the old `download_file` method in sunpy.
[here](https://github.com/vermashresth/sunpy/blob/issue2088/sunpy/util/net.py#L224:L225)
```
    if not overwrite and os.path.exists(path):
                    path = replacement_filename(path)
```
4. The file is downloaded in the directory specified in parameter instead of default cache directory (as is in the astropy version). 
4. The new function need not handle deleting of temporary files so that part is removed.

Reply:
Is there no way to either wrap the Astropy function or submit a patch to Astropy so we can use the function? 

I am against taking the Astropy function wholesale if we can avoid.

Reply:
@nabobalis Can we split `download_file` function of astropy into smaller helper functions and call them from within `download_file`. These helper functions can then be used inside our sunpy's download function too in a more clean way.

Reply:
If that could work but I am unsure how many effort that would take. You would have to ask astropy people I would reckon.

Reply:
And the wrapping option? I don't get how to proceed with that.

Reply:
Well It was more of a question than a solution.

Reply:
Okay, I'll try these options then and tell you how it goes.